src/components/AddItemForm/AddItemForm.js

```diff
@@ -36,22 +41,35 @@ const AddItemForm = ({ listId }) => {
         .collection(`lists/${listId}/items`)
         .get()
         .then(async (querySnapshot) => {
-          // create array of normalized item names from Firestore response (querySnapshot)
-          const dbItemArray = querySnapshot.docs.map((doc) =>
-            normalizeInput(doc.data().itemName),
-          );
-
-          // if item exists, show error message and put focus on field
-          // otherwise, continue with adding to database
-          if (dbItemArray.includes(normalizeInput(formValues.itemName))) {
-            setErrorMessage('Item already exists in Shopping List.');
-            itemNameRef.current.focus();
-          } else {
+          if (!querySnapshot.empty) {
+            // create array of normalized item names from Firestore response (querySnapshot)
+            const dbItemArray = querySnapshot.docs.map((doc) =>
+              normalizeInput(doc.data().itemName),
+            );
+
+            // if item exists, show error message and put focus on field
+            // otherwise, continue with adding to database
+            if (dbItemArray.includes(normalizeInput(formValues.itemName))) {
+              setItemErrorMessage('Item already exists in Shopping List.');
+              itemNameRef.current.focus();
+            } else {
+              addItemToDatabase();
+            }
+          } else if (!querySnapshot.metadata.fromCache) {
+            // if empty results and data is NOT cached, just add the item
             addItemToDatabase();
+          } else {
+            // empty results and querySnapshot.metadata.fromCache indicates a connection issue
+            setAddItemFormError(
+              'Sorry, there was a problem adding your item. Please check your connection and try again.',
+            );
           }
         });
     } catch (err) {
       console.log(err);
+      setAddItemFormError(
+        'Sorry, there was a problem adding your item. Please check your connection and try again.',
+      );
```

Comment:
For the errors that are affiliated with the default catch, is it basically safe to assume there is some problem with the user's internet? Is this the error that would get thrown when there is some problem with the database where the `get()` promise rejects? Either way, I think the actual message is probably fine, just want to clarify for my own understanding.

Reply:
I honestly can't think of any situation where the catch would run? Since a connection issue doesn't throw any error, and instead is handled in the else statement just above, any other problem would have to be something weird with Firebase which we figured is just as well reported as a vague connection issue. But actually... looking at this again, I think we could neaten it up by throwing an error in the else statement, so we only have to write setting the form error state once.